### DISCUSSION TITLE
Issue: Column Name Case Sensitivity Behavior with SQLAlchemy and cx_Oracle. Potential bug?

### DISCUSSION THREAD

**Discussion by jscov** [sqlalchemy/sqlalchemy#12154] [Usage Questions] [answered]
Dear sqlalchemy Community,

I found this problem when working with sqlalchemy and wanted to doublecheck here before reporting a bug:

### Problem Description:
When querying an Oracle database using SQLAlchemy with cx_Oracle, there is inconsistent behavior in how column names are handled in the query results. Specifically, when using pandas.read_sql() to fetch data:

- SQLAlchemy connection:

  A column without a special character (e.g., "A") is returned as "a" (lowercased).
  A column with a dash (e.g., "A-") retains its original casing.

- Direct cx_Oracle connection:

  Both column names are returned with their original casing.

### Reproduction Code:

```
import pandas
from sqlalchemy import create_engine
from sqlalchemy.pool import NullPool
import cx_Oracle

user =""
password =""
dsn = ""

pool = cx_Oracle.SessionPool(   user=user,
                                password=password,
                                dsn=dsn)

engine = create_engine("oracle+cx_oracle://", creator=pool.acquire, poolclass=NullPool)
sqlalchemy_connection = engine.connect()

sql1 = """SELECT NULL AS "A" FROM dual"""
sql2 = """SELECT NULL AS "A-" FROM dual"""

data = pandas.read_sql(sql1, sqlalchemy_connection)
print("no dash, sqlalchemy", data)
# returns column name "a" instead of "A"

data = pandas.read_sql(sql2, sqlalchemy_connection)
print("dash, sqlalchemy", data)
# returns column name "A-" as expected

connection = pool.acquire()

data = pandas.read_sql(sql1, connection)
print("no dash, cx_oracle", data)
# returns column name "A" as expected

data = pandas.read_sql(sql2, connection)
print("no dash, cx_oracle", data)
# returns column name "A-" as expected
```

### Observed Behavior:

- SQLAlchemy Connection:

  Query: SELECT NULL AS "A"
  Result: "a" (lowercase).
  Query: SELECT NULL AS "A-"
  Result: "A-" (original casing).

- Direct cx_Oracle Connection:

  Query: SELECT NULL AS "A"
  Result: "A" (original casing).
  Query: SELECT NULL AS "A-"
  Result: "A-" (original casing).

### Expected Behavior:
Column names should retain their original casing for both queries when using SQLAlchemy, consistent with cx_Oracle.

### Additional Context:
SQLAlchemy version: 2.0.36
cx_Oracle version: 8.3.0
Pandas version: 2.2.3
Python version: 3.11.10

The inconsistency in handling column names creates confusion when working with case-sensitive column aliases. It is particularly problematic when using pandas for data processing, as column name transformations can cause downstream errors or unexpected behavior.

### Suggested Resolution:
Ensure SQLAlchemy respects the original casing of column names when executing queries, aligning its behavior with cx_Oracle.

**Comment by zzzeek** [comment, marked as answer]
this is the correct behavior.  SQLAlchemy considers column identifiers as either case sensitive names or case insensitive.   Most databases including PostgreSQL, MySQL, SQL Server, SQlite will return case insensitive identifiers as lower case.  That is, these statements are all the same and both return the column name as ``myname``, all lower case on these databases (but not on Oracle):

```sql

select MyNAME from table

select myname from table

select MYNAME from table
```

case sensitive names are indicated using quotes - these statements are all different and target different column names, and will return the exact casing on those databases (but NOT on Oracle, which we'll see in a moment):

```sql

select "MyNAME" from table

select "myname" from table

select "MYNAME" from table

```

the problem with Oracle is that it indicates case insensitive names in a result in the opposite way, as all UPPERCASE.   In order to support backend-agnostic code, SQLAlchemy *normalizes* Oracle identifiers coming back in a result to be lowercase, if they are detected as being case insensitive.  The only indicator that an identifier is case insensitive is if it comes back as "ALL_UPPERCASE", including that it has no special characters like dashes which you mention (these all imply the identifier was quoted).

TL;DR Oracle has a single non-ideal case where if an identifier is specified as a quoted, all uppercase name, like this:

```sql
select "MYNAME" from table
```

SQLAlchemy cannot differentiate this from any of these statements:

```sql
select MYNAME from table
select MyName from table
select myname from table
```

(note the column name is not quoted)

there was an epic discussion about this behavior at #10781.

SQLAlchemy 2.1 includes a new execution option called [driver_column_names](https://docs.sqlalchemy.org/en/21/core/connections.html#sqlalchemy.engine.Connection.execution_options.params.driver_column_names
) which you can use to turn off the name normalize feature on a per-statement basis (or per connection, per engine).

**Reply by jscov** [reply]
Thanks a lot, very helpful! I hope that the problem will be resolved once SQLAlchemy 2.1 is published and I can set [driver_column_names](https://docs.sqlalchemy.org/en/21/core/connections.html#sqlalchemy.engine.Connection.execution_options.params.driver_column_names)!

**Comment by cjbj** [comment]
@jscov a side comment: please move from cx_Oracle to python-oracledb: https://docs.sqlalchemy.org/en/20/dialects/oracle.html#module-sqlalchemy.dialects.oracle.oracledb
